File: hw1/moon/moon.py

class Moon:

    # constructor
    def __init__(self,size : int,color : str, phase : int = 0) -> None:

        # properties...
        self._size : int = size
        self._color : str = color
        self._phase : int = phase

    def isNewMoon(self) -> bool:
        return self._phase == 0

    def cycle(self) -> None:
        self._phase = (self._phase + 1) % 4

    @property
    def phase(self):
        return self._phase

    @property
    def color(self):
        return self._color

    @property
    def size(self):
        return self._size



    

File: js-oo/docs/moon.mjs
export class Moon {
    constructor(div, id, config) {
	   this._div = div;
	   this._id = id;
	   this._size = config.size;
	   this._color = config.color;
	   this._phase = config.phase;
       this._div.innerHTML = "Moon(id=" + id + ", color="  + color + ")";
    }

    config(div,id,config) {
    }

    isNewMoon() {
        return this._phase == 0
    }

    cycle() {
        this._phase = (this._phase + 1) % 4
    }

    get phase() {
        return this._phase
    }

    get color() {
        return this._color
    }

    set color(value) {
        if (! ["white","silver"].includes(value)) {
            throw Error('invalid value')
        }
        this._color = value
    }

    get size() {
        return this._size
    }
}

export function MoonInstantiate() {
    const divs=document.getElementsByClassName("moon");
    for (let i=0; i<divs.length; ++i) {
        const div=divs[i];
        const id = "moon-" + i;
        const config = JSON.parse(div.getAttribute("data-config"));
        const moon = new Moon(div,id,config)
    }
}
